src/ProtoRoute/Message.hs

```diff
@@ -133,24 +133,31 @@ thereExistsMF (MessageField name _) = unFieldName name /= ""
 validateMessage :: MessageSchema -> Message -> Either ValidationError ()
 validateMessage (MessageSchema sn sFields) (Message mn mFields) = do
     verify (mn /= sn) incorrectMsgNames
-    let nameFieldMap = groupedFields
-    let results = DMS.elems nameFieldMap
-    if null (catThese results)
-    then Left invalid
-    else Right ()
+    verify (null $ catThese $ M.elems groupedFields) invalid
   where
-    groupedFields :: DMS.Map FieldName (These SchemaField MessageField)
-    groupedFields = DMS.fromList $ zip fieldsNames theseFields
-
-    fieldsNames =
-        map schemaFieldName sFields `union` map messageFieldName mFields
-    theseFields = zipWith (curry toThese) sFields mFields
-    toThese :: (SchemaField, MessageField) -> These SchemaField MessageField
-    toThese (sf, mf) =
-      if thereExistsSF sf && thereExistsMF mf
-      then These sf mf
-      else if | thereExistsSF sf && not (thereExistsMF mf) -> This sf
-              | not (thereExistsSF sf) && thereExistsMF mf -> That mf
+    groupedFields :: M.Map FieldName (These SchemaField MessageField)
+    groupedFields = unionThese schemaMap msgMap
+
+    unionThese :: Ord k => M.Map k a -> M.Map k b -> M.Map k (These a b)
+    unionThese mapA mapB = sortThemOut (M.assocs mapA) (M.assocs mapB)
+
+    fieldNames :: [FieldName]
+    fieldNames = map schemaFieldName sFields `union` map messageFieldName mFields
```

Comment:
`union` of field names doesn't make much sense. You don't need to use same `fieldNames` for `schemaMap` and `msgMap`. You need to use `FieldName` from the given list.